File: Output/output_Copilot/c/Hard/1865-stickers-to-spell-word.c
/**https://leetcode.com/problems/stickers-to-spell-word/ */
//We are given n different types of stickers. Each sticker has a lowercase English word on it.
//You would like to spell out the given string target by cutting individual letters from your collection of stickers and rearranging them. You can use each sticker more than once if you want, and you have infinite quantities of each sticker.
//Return the minimum number of stickers that you need to spell out target. If the task is impossible, return -1.
//Note: In all test cases, all words were chosen randomly from the 1000 most common US English words, and target was chosen as a concatenation of two random words.
// 
//Example 1:
//Input: stickers = ["with","example","science"], target = "thehat"
//Output: 3
//Explanation:
//We can use 2 "with" stickers, and 1 "example" sticker.
//After cutting and rearrange the letters of those stickers, we can form the target "thehat".
//Also, this is the minimum number of stickers necessary to form the target string.
//Example 2:
//Input: stickers = ["notice","possible"], target = "basicbasic"
//Output: -1
//Explanation:
//We cannot form the target "basicbasic" from cutting letters from the given stickers.
// 
//Constraints:
//	n == stickers.length
//	1 <= n <= 50
//	1 <= stickers[i].length <= 10
//	1 <= target.length <= 15


        int minStickers(char** stickers, int stickersSize, char* target) {
            int targetLen = strlen(target);
            int targetMask = (1 << targetLen) - 1;
            int* dp = (int*)malloc((1 << targetLen) * sizeof(int));
            for (int i = 0; i < (1 << targetLen); i++) {
                dp[i] = INT_MAX;
            }
            dp[0] = 0;

            for (int state = 0; state < (1 << targetLen); state++) {
                if (dp[state] == INT_MAX) continue;
                for (int i = 0; i < stickersSize; i++) {
                    int nextState = state;
                    int stickerLen = strlen(stickers[i]);
                    for (int j = 0; j < stickerLen; j++) {
                        for (int k = 0; k < targetLen; k++) {
                            if (target[k] == stickers[i][j] && !(nextState & (1 << k))) {
                                nextState |= (1 << k);
                                break;
                            }
                        }
                    }
                    dp[nextState] = dp[nextState] < dp[state] + 1 ? dp[nextState] : dp[state] + 1;
                }
            }

            int result = dp[targetMask] == INT_MAX ? -1 : dp[targetMask];
            free(dp);
            return result;
        }

            

File: Output/output_Copilot/python/Hard/1865-stickers-to-spell-word.py
#https://leetcode.com/problems/stickers-to-spell-word/
#We are given n different types of stickers. Each sticker has a lowercase English word on it.
#You would like to spell out the given string target by cutting individual letters from your collection of stickers and rearranging them. You can use each sticker more than once if you want, and you have infinite quantities of each sticker.
#Return the minimum number of stickers that you need to spell out target. If the task is impossible, return -1.
#Note: In all test cases, all words were chosen randomly from the 1000 most common US English words, and target was chosen as a concatenation of two random words.
# 
#Example 1:
#Input: stickers = ["with","example","science"], target = "thehat"
#Output: 3
#Explanation:
#We can use 2 "with" stickers, and 1 "example" sticker.
#After cutting and rearrange the letters of those stickers, we can form the target "thehat".
#Also, this is the minimum number of stickers necessary to form the target string.
#Example 2:
#Input: stickers = ["notice","possible"], target = "basicbasic"
#Output: -1
#Explanation:
#We cannot form the target "basicbasic" from cutting letters from the given stickers.
# 
#Constraints:
#	n == stickers.length
#	1 <= n <= 50
#	1 <= stickers[i].length <= 10
#	1 <= target.length <= 15
#	stickers[i] and target consist of lowercase English letters.
class Solution(object):
    def minStickers(self, stickers, target):
        """
        :type stickers: List[str]
        :type target: str
        :rtype: int
        """
        dp = [float('inf')] * (1 << len(target))
        dp[0] = 0
        for sticker in stickers:
            count = [0] * 26
            for ch in sticker:
                count[ord(ch) - ord('a')] += 1
            for i in range(1 << len(target)):
                if dp[i] == float('inf'):
                    continue
                j = i
                for k in range(len(target)):
                    if j & (1 << k) == 0 and count[ord(target[k]) - ord('a')] > 0:
                        j |= (1 << k)
                        count[ord(target[k]) - ord('a')] -= 1
                dp[j] = min(dp[j], dp[i] + 1)
        return dp[-1] if dp[-1] != float('inf') else -1        